include/flatbuffers/util.h

```diff
@@ -170,28 +171,164 @@ inline std::string IntToStringHex(int i, int xdigits) {
   // clang-format on
 }
 
-// Portable implementation of strtoll().
-inline int64_t StringToInt(const char *str, char **endptr = nullptr,
+// Adaptor for strtoull()/strtoll().
+// Flatbuffers accepts numbers with any count of leading zeros (-009 is -9),
+// while strtoll with base=0 interprets first leading zero as octal prefix.
+// In future, it is possible to add prefixed 0b0101.
+// The strtoll returns overflow state using errno code (thread-local var).
+// 1) By request it checks errno code (out of range).
+// To request errno check: if (endptr && *endptr == str) is true.
+// 2) If base <= 0, function try to detect base of number by prefix.
+//
+// Return value (like strtoull and strtoll, but reject partial result):
+// - If successful, an integer value corresponding to the str is returned.
+// - If full string conversion can't be performed, ​0​ is returned.
+// - If the converted value falls out of range of corresponding return type, a
+// range error occurs. In this case value MAX(T)/MIN(T) is returned.
+// If errno check is requested then the endptr will
+// point to the start of input string (*endptr=str).
+template<typename T>
+inline T StringToInteger64Impl(const char *const str, const char **endptr,
```

Comment:
these are pretty big templates, to what extend can we factor some of this out to go into the .cpp? If it becomes too messy then ignore this comment.